### ISSUE TITLE
Parallel and timed apply

### ISSUE THREAD

**Issue by marcglobality** [pdpipe/pdpipe#40]
First of all, thanks for this amazing library, really useful. Now going to the issue / request:

It seems to me that ApplyByCols and ApplyToRows could have a parameter to make them run in parallel. Has this ever been  considered as a feature? I think it could speed up pipelines quite a lot, esp. useful for big dataframes. WDYT?

Also, I see that pipeline.fit and pipeline.transform have the `timed` bool. Would it be possible to add the same for apply? I know I can do pipeline.transform(...., time=True), but don't see a reason why apply cannot have it

**Comment by shaypal5** [comment]
1. Hmm. Never considered this. I will definitely accept PRs around parallelization! :)
2. Definitely. Perhaps it was just an oversight. This should be an easy PR to make. Perhaps you can write it?

Also, would you mind separating this into two distinct issues?
The first one is complex, while the other I can label with the `first good issue` label. :)

**Comment by yarkhinephyo** [comment]
Hi! I would like to make a new issue for the No.2 and work on it if its okay :)

**Comment by shaypal5** [comment]
Yeah, definitely! Go fo it and let me know how it goes! :)

**Comment by jbbqqf** [comment]
Hi @shaypal5! Triaging older issues this week — proposing to split this one.

Two distinct asks were entangled here:

1. **Parallelize `ApplyByCols` / `ApplyToRows`** — your 2020-10-20 reply called this complex; you suggested splitting it from sub-issue 2. It is **still unimplemented**: `rg -n "n_jobs|joblib|multiprocessing|parallel" src/pdpipe/col_generation.py` returns nothing in the current `master` (`3140064`). If you still want this, it deserves a fresh issue with a concrete design (joblib? multiprocessing? per-stage `n_jobs=` like sklearn?) so contributors can scope it.

2. **Add `time=True` to `apply()`** — this **has shipped**. [`src/pdpipe/core.py:1464-1534`](https://github.com/pdpipe/pdpipe/blob/master/src/pdpipe/core.py#L1464-L1534): `PdPipelineStage.apply(self, X, y=None, exraise=None, verbose=False, time=False, ...)` now accepts and forwards `time` to `__timed_fit_transform` / `__timed_transform`.

Quick check:

```python
import pandas as pd, pdpipe as pdp
df = pd.DataFrame({'a': [1, 2, 3]})
pdp.ColDrop('a').apply(df, time=True)  # works in current main, prints timing
```

Suggest closing this issue and (if you want to keep the parallelization track alive) opening a fresh, focused one. Happy to file the new issue myself with a sklearn-style `n_jobs=` proposal if that helps.

(Disclosure: I drafted this comment with help from Claude Code while triaging stale issues; the references above were verified manually against current `master`.)

**Comment by shaypal5** [comment]
Closing after splitting the remaining work. The apply(time=True) part has shipped; the still-open parallelization request deserves its own focused issue with design constraints and acceptance criteria.
